OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -0,0 +1,429 @@
+					throw new InvalidDataException("Empty parenthesis at index {0}".F(tokens[i].Index));
+
+				// A variable or number must be followed by a binary operation or by a closing parenthesis
+				if (tokens[i].Arity == 0 && !(tokens[i + 1].Arity > 1 || tokens[i + 1].Closer))
+					throw new InvalidDataException("Missing binary operation at index {0}".F(tokens[i + 1].Index));
+			}
+
+			// Expressions can't start with a binary operation
+			if (tokens[0].Arity > 1)
+				throw new InvalidDataException("Unexpected token `{0}` at index `{1}`".F(tokens[0].Symbol, tokens[0].Index));
+
+			// Expressions can't end with a binary or unary operation
+			if (tokens[tokens.Count - 1].Arity > 0)
+				throw new InvalidDataException("Unexpected token `{0}` at index `{1}`".F(tokens[tokens.Count - 1].Symbol, tokens[tokens.Count - 1].Index));
+
+			// Binary operations must be preceeded by a closing paren or a variable
+			// Binary operations must be followed by an opening paren, a variable, or a unary operation
+			for (var i = 1; i < tokens.Count - 1; i++)
+			{
+				if (tokens[i].Arity > 1 && (
+					!(tokens[i - 1].Closer || tokens[i - 1].Arity == 0) ||
+					!(tokens[i + 1].Opener || tokens[i + 1].Arity == 0 || tokens[i + 1].Arity == 1)))
+					throw new InvalidDataException("Unexpected token `{0}` at index `{1}`".F(tokens[i].Symbol, tokens[i].Index));
+			}
+
+			// Convert to postfix (discarding parentheses) ready for evaluation
+			var postfix = new ReadOnlyList<Token>(ToPostfix(tokens).ToList());
+			var types = new Stack<ValueType>();
+			evaluators = postfix.SelectMany((token, index) => token.GetEvaluators(types, postfix, index)).ToArray();
+		}
+
+		enum CharClass { Whitespace, Operator, Mixed, Id, Digit }
+
+		static CharClass CharClassOf(char c)
+		{
+			switch (c)
+			{
+				case '~':
+				case '!':
+				case '%':
+				case '^':
+				case '&':
+				case '*':
+				case '(':
+				case ')':
+				case '+':
+				case '=':
+				case '[':
+				case ']':
+				case '{':
+				case '}':
+				case '|':
+				case ':':
+				case ';':
+				case '\'':
+				case '"':
+				case '<':
+				case '>':
+				case '?':
+				case ',':
+				case '/':
+					return CharClass.Operator;
+
+				case '.':
+				case '@':
+				case '$':
+				case '-':
+					return CharClass.Mixed;
+
+				case '0':
+				case '1':
+				case '2':
+				case '3':
+				case '4':
+				case '5':
+				case '6':
+				case '7':
+				case '8':
+				case '9':
+					return CharClass.Digit;
+
+				// Fast-track normal whitespace
+				case ' ':
+				case '\t':
+				case '\n':
+				case '\r':
+					return CharClass.Whitespace;
+
+				// Should other whitespace be tested?
+				default:
+					return char.IsWhiteSpace(c) ? CharClass.Whitespace : CharClass.Id;
+			}
+		}
+
+		public enum Keyword { NotAKeyword, True, False, And, Or }
+
+		public static Keyword GetKeyword(string expression)
+		{
+			switch (expression[0])
+			{
+				case 'a':
+					return expression == "and" ? Keyword.And : Keyword.NotAKeyword;
```

Comment:
Please lets not support multiple aliases of the same operator.  It has taken years to force some consistency into our boolean yaml definitions.